src/vidpf.rs

```diff
@@ -0,0 +1,822 @@
+
+    /// Panics if weight lengths are different.
+    fn add(self, rhs: Self) -> Self::Output {
+        assert_eq!(
+            self.0.len(),
+            rhs.0.len(),
+            "{}",
+            VidpfError::InvalidWeightLength
+        );
+
+        Self(zip(self.0, rhs.0).map(|(a, b)| a.add(b)).collect())
+    }
+}
+
+impl<F: FieldElement> AddAssign for VidpfWeight<F> {
+    /// Panics if weight lengths are different.
+    fn add_assign(&mut self, rhs: Self) {
+        assert_eq!(
+            self.0.len(),
+            rhs.0.len(),
+            "{}",
+            VidpfError::InvalidWeightLength
+        );
+
+        zip(&mut self.0, rhs.0).for_each(|(a, b)| a.add_assign(b));
+    }
+}
+
+impl<F: FieldElement> Sub for VidpfWeight<F> {
+    type Output = Self;
+
+    /// Panics if weight lengths are different.
+    fn sub(self, rhs: Self) -> Self::Output {
+        assert_eq!(
+            self.0.len(),
+            rhs.0.len(),
+            "{}",
+            VidpfError::InvalidWeightLength
+        );
+
+        Self(zip(self.0, rhs.0).map(|(a, b)| a.sub(b)).collect())
+    }
+}
+
+impl<F: FieldElement> Encode for VidpfWeight<F> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        for e in &self.0 {
+            F::encode(e, bytes)?;
+        }
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.0.len() * F::ENCODED_SIZE)
+    }
+}
+
+impl<F: FieldElement> ParameterizedDecode<<Self as IdpfValue>::ValueParameter> for VidpfWeight<F> {
+    fn decode_with_param(
+        decoding_parameter: &<Self as IdpfValue>::ValueParameter,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        let mut v = Vec::with_capacity(*decoding_parameter);
+        for _ in 0..*decoding_parameter {
+            v.push(F::decode_with_param(&(), bytes)?);
+        }
+
+        Ok(Self(v))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::field::Field128;
+
+    use super::VidpfWeight;
+
+    type TestWeight = VidpfWeight<Field128>;
+    const TEST_WEIGHT_LEN: usize = 3;
+    const TEST_NONCE_SIZE: usize = 16;
+    const TEST_NONCE: &[u8; TEST_NONCE_SIZE] = b"Test Nonce VIDPF";
+
+    mod vidpf {
+        use crate::{
+            idpf::IdpfValue,
+            vidpf::{Vidpf, VidpfEvalState, VidpfInput, VidpfKey, VidpfPublicShare, VidpfServerId},
+        };
+
+        use super::{TestWeight, TEST_NONCE, TEST_NONCE_SIZE, TEST_WEIGHT_LEN};
+
+        fn vidpf_gen_setup(
+            input: &VidpfInput,
+            weight: &TestWeight,
+        ) -> (
+            Vidpf<TestWeight, TEST_NONCE_SIZE>,
+            VidpfPublicShare<TestWeight>,
+            [VidpfKey; 2],
+            [u8; TEST_NONCE_SIZE],
+        ) {
+            let vidpf = Vidpf::new(TEST_WEIGHT_LEN);
+            let (public, keys) = vidpf.gen(input, weight, TEST_NONCE).unwrap();
+            (vidpf, public, keys, *TEST_NONCE)
+        }
+
+        #[test]
+        #[should_panic]
```

Comment:
Since this panic is coming from an unwrap in test code, we could improve this by instead writing `let err = vidpf.gen_with_keys(...).unwrap_err(); assert_matches!(err, Error::SameKeyId);`. This would prevent false negatives if the test panicked for some other reason.

Reply:
I adopted this idea, however `assert_matches` is in nightly.

Reply:
We have a dev dependency on the `assert_matches` crate, which provides such a macro. It's nice that it's getting uplifted into std!